src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,324 @@
+
+      cs[i] = c;
+    }
+
+    if (!weights || iters) {
+      double cur_ex_nodes = 0;
+      for (auto it : hop_map)
+        cur_ex_nodes += std::min((FloatType)1, it.second);
+      if (cur_ex_nodes / prev_ex_nodes >= 1 - eps)
+        break;
+      prev_ex_nodes = cur_ex_nodes;
+    }
+  }
+}
+
+// Template for picking non-zero values row-wise.
+template <typename IdxType, typename FloatType>
+std::pair<COOMatrix, FloatArray> CSRLaborPick(
+    CSRMatrix mat,
+    IdArray rows,
+    int64_t num_picks,
+    FloatArray prob,
+    int importance_sampling,
+    IdArray random_seed_arr,
+    IdArray NIDs) {
+  using namespace aten;
+  const IdxType* indptr = mat.indptr.Ptr<IdxType>();
+  const IdxType* indices = mat.indices.Ptr<IdxType>();
+  const IdxType* data = CSRHasData(mat) ? mat.data.Ptr<IdxType>() : nullptr;
+  const IdxType* rows_data = rows.Ptr<IdxType>();
+  const IdxType* nids = IsNullArray(NIDs) ? nullptr : NIDs.Ptr<IdxType>();
+  const auto num_rows = rows->shape[0];
+  const auto& ctx = mat.indptr->ctx;
+
+  const bool weights = !IsNullArray(prob);
+  // O(1) c computation not possible, so one more iteration is needed.
+  if (importance_sampling >= 0)
+    importance_sampling += weights;
+  // A stands for A in arXiv:2210.13339, that is edge weights
+  auto A_arr = prob;
+  FloatType *A = A_arr.Ptr<FloatType>();
+  constexpr FloatType ONE = 1;
+
+  constexpr auto fidtype = DGLDataTypeTraits<FloatType>::dtype;
+
+  // cs stands for c_s in arXiv:2210.13339
+  FloatArray cs_array = NDArray::Empty({num_rows}, fidtype, ctx);
+  FloatType* cs = cs_array.Ptr<FloatType>();
+  // ds stands for A_{*s} in arXiv:2210.13339
+  FloatArray ds_array = NDArray::Empty({num_rows}, fidtype, ctx);
+  FloatType* ds = ds_array.Ptr<FloatType>();
+
+  IdxType max_degree = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_degree = indptr[rid + 1] - indptr[rid];
+    max_degree = std::max(act_degree, max_degree);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_degree;
+    // O(1) c computation, samples more than needed for weighted case, mentioned in the sentence
+    // between (10) and (11) in arXiv:2210.13339
+    cs[i] = num_picks / d;
+    ds[i] = d;
+    hop_size += act_degree;
+  }
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  if (importance_sampling)
+    compute_importance_sampling_probabilities<IdxType, FloatType>(
+        ctx,
+        fidtype,
+        max_degree,
+        num_rows,
+        importance_sampling,
+        weights,
+        rows_data,
+        indptr,
+        A,
+        indices,
+        (IdxType)num_picks,
+        ds,
+        cs,
+        hop_map);
+
+  const uint64_t random_seed = IsNullArray(random_seed_arr) ?
+      RandomEngine::ThreadLocal()->RandInt(1000000000) : random_seed_arr.Ptr<int64_t>()[0];
```

Comment:
It is a tensor because the tensor can be empty in case the user doesn't pass anything. Then the C++ code can check if the tensor is empty and if it is, then can get the random seed from DGL PRNG. So, I am using the tensor like std::optional.

Reply:
And also, if the python side doesn't have that logic, then I can't ensure that the same random_seed is used for different edge types.

Reply:
This logic exists so that when anyone uses the c++ api, they don't have to pass anything for random seed, it will be handled automatically. Should improve c++ api user experience.

Reply:
OK.  You can keep it for now.